capsules/src/net/thread/tlv.rs

```diff
@@ -0,0 +1,1411 @@
+//! only when a variable-length value must be decoded from network byte
+//! order before it can be interpreted correctly. Excluded from this case
+//! are variable-length values that contain data that must later be
+//! decoded by the caller before being interpreted (for example,
+//! sub-TLVs). Such a value is instead returned as a slice of the original
+//! buffer passed to the decode function.
+//!
+//!
+//! Author: Mateo Garcia
+//!         mateog@stanford.edu
+
+
+// TODO: Move the MLE explanation above to the MLE module, when it is created.
+
+
+// NOTES FOR DEBUGGING:
+// - .to_be() may not have been called on values wider than one byte
+// - encode_bytes_be may have been used instead of encode_bytes
+// - decode_bytes_be may have been used instead of decode_bytes
+// - See 4.5.25 Active Operational Dataset TLV and 4.5.26 Pending Operational Dataset TLV
+//    - Are Active and Pending Timestamp TLVs, respectively, required to be sent as well
+//      if either of the dataset tlvs are sent?
+
+
+use core::mem;
+use net::stream::{decode_u8, decode_u16, decode_u32, decode_bytes_be};
+use net::stream::{encode_u8, encode_u16, encode_u32, encode_bytes, encode_bytes_be};
+use net::stream::SResult;
+
+const TL_WIDTH: usize = 2; // Type and length fields of TLV are each one byte.
+const MAX_VALUE_LENGTH: usize = 128; // Assume a TLV value will be no longer than 128 bytes.
+
+/// Type-Length-Value structure.
+pub enum Tlv<'a> {
+    SourceAddress(u16),
+    Mode(u8),
+    Timeout(u32),
+    Challenge([u8; 8]), // Byte string max length 8 bytes.
+    Response([u8; 8]), // Byte string max length 8 bytes.
+    LinkLayerFrameCounter(u32),
+    // LinkQuality,                  // TLV type Not used in Thread
+    // NetworkParameter,             // TLV type Not used in Thread
+    MleFrameCounter(u32),
+    /*
+    TODO: Not required to implement MLE for SED
+    Route64,
+    */
+    Address16(u16),
+    LeaderData {
+        partition_id: u32,
+        weighting: u8,
+        data_version: u8,
+        stable_data_version: u8,
+        leader_router_id: u8,
+    },
+    NetworkData(&'a [u8]),
+    TlvRequest(&'a [u8]),
+    ScanMask(u8),
+    Connectivity {
+        parent_priority: u8,
+        link_quality_3: u8,
+        link_quality_2: u8,
+        link_quality_1: u8,
+        leader_cost: u8,
+        id_sequence: u8,
+        active_routers: u8,
+        sed_buffer_size: Option<u16>,
+        sed_datagram_count: Option<u8>,
+    },
+    LinkMargin(u8),
+    Status(u8),
+    Version(u16),
+    /*
+    TODO: Not required to implement MLE for SED
+    AddressRegistration
+    AddressRegistration
+    Channel
+    PanId
+    ActiveTimestamp
+    PendingTimestamp
+    ThreadDiscovery
+    */
+    ActiveOperationalDataset(&'a [u8]),
+    PendingOperationalDataset(&'a [u8]),
+}
+
+impl<'a> Tlv<'a> {
+    /// Serializes TLV data in `buf` into the format specific to the TLV
+    /// type.
```

Comment:
I'm not sure we can answer that question just yet..

128 is an arbitrary number that I chose as the maximum length that a single field in a TLV value can have. It doesn't have to do with `encode` — it's important for `decode`, in cases where I have to decode a TLV value that can vary in length and stick it into a field in a `Tlv` variant to return to the caller of the decode function. For example, look at the `NetworkDataTlv::CommissioningData` variant. When I'm [decoding](https://github.com/ptcrews/tock/blob/31e8b12cf6eda86421dda39fc1bd0fee06591cab/capsules/src/net/thread/tlv.rs#L645) this type of TLV, according to the spec, the `com_data` field can vary in length, so ideally I'd want `com_data` to be a dynamically sized array. The way I would normally deal with that is by making `com_data` a reference to a slice of the buffer passed to `decode`. But because the contents of `com_data` must be decoded from network byte order before being returned to the user, I can't just return a slice of the original buffer. The solution I came up with was making the `com_data` field an array with a fixed length of 128, which I'm guessing will be sufficiently large to store the decoded contents.

That's where 128 comes into play in this module.

The buffer passed to `encode` must simply be large enough to store all of the fields of a given `Tlv` variant (which together represent the TLV value), plus two bytes for the TLV type and the TLV length. I was assuming that this buffer would simply be a slice of the buffer being used to encode the entire MLE of which this particular TLV is a part. It's unclear how large that MLE buffer should be; it's something we'll have to figure out when we implement MLE.